src/net_processing.cpp

```diff
@@ -1003,7 +1003,7 @@ void PeerManagerImpl::FinalizeNode(const CNode& node, bool& fUpdateConnectionTim
     for (const QueuedBlock& entry : state->vBlocksInFlight) {
         mapBlocksInFlight.erase(entry.hash);
     }
-    EraseOrphansFor(nodeid);
+    WITH_LOCK(g_cs_orphans, EraseOrphansFor(nodeid));
```

Comment:
denialofservices_tests locks `g_cs_orphans` then calls `EraseOrphansFor`, so it would have required more changes to have `EraseOrphansFor` do the locking.

I don't think `FinalizeNode` is called from anywhere that holds `g_cs_orphans`? Am I missing somewhere?

Reply:
Here: https://github.com/bitcoin/bitcoin/blob/b9f41df1ead4b6a83a51fc41966b111c8459c313/src/init.cpp#L213-L214 ?

Reply:
Yes, that's it!

StopNodes() calls DeleteNode() calls FinalizeNode() calls EraseForPeer(); with EraseForPeer needing `g_cs_orphans`. StopNodes locks `cs_vNodes` before calling DeleteNode, so `g_cs_orphans` has to be locked before calling CConnman::StopNodes but only if it's linked to a PeerManagerImpl.

StopNodes is also called from `~CConnman` so that path will still get a lock order violation error, I think.

DeleteNode is also called from DisconnectNodes which is called from ThreadSocketHandler; but there it's called after `cs_vNodes` is released, so lock ordering is not an issue, but `g_cs_orphans` still has to get locked prior to doing the work in EraseForPeer, so responsibility can't be moved all the way back up the chain. 

So I don't think there's a simple way to make it non-recursive for this PR; for a future PR, I think the aim is to make `g_cs_orphans` become `private: TxOrphanage::m_mutex` and not have it be the last thing locked / first thing released so that the ordering constraint goes away, and then it should trivially be non-recursive.

Reply:
> for a future PR, I think the aim is to make g_cs_orphans become private: TxOrphanage::m_mutex and not have it be the last thing locked / first thing released so that the ordering constraint goes away, and then it should trivially be non-recursive.

Concept ACK :)

Reply:
https://github.com/ajtowns/bitcoin/commits/202102-orphanworkset has work-in-progress that turns g_cs_orphans a protected m_mutex, doesn't quite turn orphans back into the responsibility of who sent them rather than who sent their final parent though.